nixos/modules/services/development/gemstash.nix

```diff
@@ -0,0 +1,158 @@
+      type = types.nullOr types.attrs;
+      default = null;
+      description = "Additional Sequel.connect options to use";
+    };
+    rubyGemsUrl = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      description = "The default gem source URL";
+    };
+    ignoreGemfileSource = mkOption {
+      type = types.nullOr types.bool;
+      default = null;
+      description = "Ignore the source specified in Gemfile and always use :rubygems_url as gems upstream";
+    };
+    pumaThreads = mkOption {
+      type = types.nullOr types.int;
+      default = null;
+      description = "The number of threads used for the Gemstash server";
+    };
+    bind = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      description = "The binding used to start the Gemstash server";
+    };
+    protectedFetch = mkOption {
+      type = types.nullOr types.bool;
+      default = null;
+      description = "Tells Gemstash to authenticate via an API key before allowing the fetching of private gems and specs";
+    };
+    fetchTimeout = mkOption {
+      type = types.nullOr types.int;
+      default = null;
+      description = "The timeout setting for fetching gems";
+    };
+    logFile = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      description = "The name of the file to use for logging";
+    };
+
+    extraOptions = mkOption {
+      type = types.attrs;
+      default = { };
+      description = "Extra options for the config file";
+    };
+  };
+
+  config =
+    let
+      cfg = config.services.gemstash;
+      gemstash_settings_names = {
+        basePath = "base_path";
+        cacheType = "cache_type";
+        memcachedServers = "memcached_servers";
+        dbAdapter = "db_adapter";
+        dbUrl = "db_url";
+        dbConnectionOptions = "db_connection_options";
+        rubyGemsUrl = "ruby_gems_url";
+        ignoreGemfileSource = "ignore_gemfile_source";
+        pumaThreads = "puma_theads";
+        bind = "bind";
+        protectedFetch = "protected_fetch";
+        fetchTimeout = "fetch_timeout";
+        logFile = "log_file";
+      };
+      gemstash_settings = listToAttrs (filter (e: e.value != null) (
+        map (name: { name = gemstash_settings_names."${name}"; value = cfg."${name}"; })
+          (attrNames gemstash_settings_names))
+      );
+      gemstash_config =
+        if cfg.configFile == null
+        then settingsFormat.generate "gemstash.yaml" (prefixColon gemstash_settings // cfg.extraOptions)
+        else cfg.configFile;
+    in
+    mkIf cfg.enable {
+      users = {
+        users.gemstash = {
+          group = "gemstash";
+          isSystemUser = true;
+        };
+        groups.gemstash = { };
+      };
+
+      systemd.services.gemstash = {
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        serviceConfig = {
+          ExecStartPre = if cfg.basePath != null
```

Comment:
`mkIf` would be better than `if`/`else` here... but generally you shouldn't do this.  Use `StateDirectory` or `systemd.tmpfiles`.

See [this](https://github.com/NixOS/nixpkgs/pull/178154) PR for an example of how to effectively use `StateDirectory`, which is the best solution to provisioning your directory here.